Source/FluentSpreadsheets/Sheets/VisitorImplementations/SheetSegmentVisitor.cs

```diff
@@ -0,0 +1,174 @@
+        IList<IList<IComponent>> rowComponents = rowVisitor.Components;
+        IList<IComponent> footerComponents = footerVisitor.Components;
+
+        ScalePrototypedComponents(data, headerComponents, rowComponents, footerComponents);
+
+        IComponent? header = null;
+        IList<IComponent> rows = Array.Empty<IComponent>();
+        IComponent? footer = null;
+
+        if (headerComponents.Count is not 0)
+            header = new HStackComponent(headerComponents);
+
+        if (rowComponents.Count is not 0)
+        {
+            var convertedRows = ConvertRowsFromSegmentGroupingToRowGrouping(rowComponents, _rowData.Count);
+
+            rows = convertedRows
+                .Select(x => (IComponent)new HStackComponent(x))
+                .ToArray();
+        }
+
+        if (footerComponents.Count is not 0)
+            footer = new HStackComponent(footerComponents);
+
+        _segment = new SheetSegmentModel(header, rows, footer);
+    }
+
+    private SheetSegmentModel Visit(ISheetSegment builder)
+    {
+        var headerVisitor = new HeaderSegmentVisitor<THeaderData>(_headerData);
+        var rowVisitor = new RowSegmentVisitor<THeaderData, TRowData>(_headerData, _rowData);
+        var footerVisitor = new FooterSegmentVisitor<THeaderData, TFooterData>(_headerData, _footerData);
+
+        builder.Accept(headerVisitor);
+        builder.Accept(rowVisitor);
+        builder.Accept(footerVisitor);
+
+        var header = headerVisitor.Component;
+        var row = rowVisitor.Components;
+        var footer = footerVisitor.Component;
+
+        return new SheetSegmentModel(header, row, footer);
+    }
+
+    private static void ScalePrototypedComponents<TDestination>(
+        IReadOnlyCollection<TDestination> data,
+        IList<IComponent> headerComponents,
+        IList<IList<IComponent>> rowComponents,
+        IList<IComponent> footerComponents)
+    {
+        for (var i = 0; i < data.Count; i++)
+        {
+            var rowWidth = CountRowWidth(i, headerComponents, rowComponents, footerComponents);
+
+            if (headerComponents.Count is not 0)
+                headerComponents[i] = ScaledComponent(headerComponents[i], rowWidth);
+
+            if (rowComponents.Count is not 0)
+            {
+                IList<IComponent>? segment = rowComponents[i];
+
+                for (var j = 0; j < segment.Count; j++)
+                {
+                    segment[j] = ScaledComponent(segment[j], rowWidth);
+                }
+            }
+
+            if (footerComponents.Count is not 0)
+                footerComponents[i] = ScaledComponent(footerComponents[i], rowWidth);
+        }
+    }
+
+    private static IComponent ScaledComponent(IComponent component, int width)
+    {
+        var scaleFactor = width / component.Size.Width;
+        return scaleFactor is 1 ? component : new ScaledComponent(component, new Scale(scaleFactor, 1));
```

Comment:
Its not really readable, maybe add something like Scale.Horizontal(ScaleAmount horizontalScale) => new Scale(horizontalScale, 1)

Reply:
There is already an extension for it, that I forgot about at the time of writing this code